src/object.d

```diff
@@ -4881,3 +4881,45 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    auto nbytes = from.length * TFrom.sizeof;
+    assert((nbytes % TTo.sizeof) == 0, "array cast misalignment");
+
+    struct Array
+    {
+        size_t length;
+        void* ptr;
+    }
+
+    auto a = cast(Array*)&from;
+    a.length = nbytes / TTo.sizeof; // jam new length
+    return *cast(TTo[]*)a;
+}
+
+unittest
+{
+    byte[] b;
+    int[] i;
+    short[] s;
+
+    b.length = 12;
+
+    i = __ArrayCast!(byte, int)(b);
+    assert(i.length == 3);
+
+    s = __ArrayCast!(byte, short)(b);
+    assert(s.length == 6);
+
+    s = __ArrayCast!(int, short)(i);
+    assert(s.length == 6);
```

Comment:
I'd also do an array comparison, or at the very least, a first-element comparison in addition to this length check